### PR TITLE
Upgrade webpack to version 5

### LINKED ISSUE #5244: Upgrade to webpack 5 and enable HMR
The goals of this issue are:

- [ ] upgrade to webpack 5
- [ ] enable HMR

### PR DESCRIPTION
**Pull request by thealexandrelara**
The current version of Webpack in this project is version 4. The goal is to upgrade to Webpack 5 to get improved build performance, access to new features and improvements, improved security, and compatibility with JavaScript. Once the upgrade is complete, and we've thoroughly tested the changes to our build system to ensure they don't cause any regressions or break existing functionality, we can enable HRM.

Fixes #5244 

#### Scenario 1: Build
1. Run `npm run build` and make sure the build finishes correctly;
2. Run `npm start` and check if the application is running without any error;

#### Scenario 2: Blocks
1. Log in to your WordPress dashboard.
2. Add a new post/page;
3. Add multiple blocks to the post/page;
4. Make sure you can edit the blocks settings and styles;
5. Click on the Save button;
6. Visit the post/page on the frontend and make sure everything is working.

#### Scenario 3: Templates
1. Log in to your WordPress dashboard.
2. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Two," "Twenty-twenty Three," etc.
3. On the left-hand side menu, click on Appearance > Editor > Templates
4. Select and click on a template from the list to start editing it. If it is a Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before;
5. Make sure everything is working.
6. Add multiple blocks to the template;
7. Make sure you can edit the blocks settings and styles;
8. Click on the Save button;
9. Visit the template page on the frontend and make sure everything is working.
10. Repeat steps 4-9 to test all the available templates.

#### Scenario 4: Testing with generated zip file in a hosted website
1. Run `npm run package-plugin:deploy`. This command will generate a zip file in the root folder of your project;
2. Create a new WooCommerce on [Pressable](https://pressable.com/) or any other hosting of your preference;
3. From your WordPress dashboard, go to Plugins > Add new. Upload the generated zip and activate the plugin;
4. Repeat the steps from the scenarios 2 and 3 above.

1. Run `npm run package-plugin:deploy`. This command will generate a zip file in the root folder of your project;
2. Repeat the steps from the scenarios 2 and 3, above.
3. Considering that you have a site with a previous version of the plugin installed (a version that still uses webpack 4). Go to Plugins > Installed plugins. Disable the current plugin version.
4. Upload the generated zip and activate the plugin;
5. Make sure that the blocks and templates are working both on the editor and on the frontend.

### Changelog

> Upgrade Webpack from version 4 to version 5

### CODE REVIEW

**Comment by gigitux** [comment]
Thanks for working on this! 
Do you think that it is fine having `${number}-frontend`? I remember that I tried to fix it without success.
On the PHP side, we register scripts with the structure `wc-${block_name}-block-frontend`.

image

File: package.json
Comment (Aljullu) [inline package.json]: If I remember correctly, we use fixed versions in `package.json` (_aka_, without the `^` caret operator), because renovatebot already takes care of updating the dependencies.

**Review by Aljullu — COMMENTED** [review]
Really exciting to see the bundle size improvements webpack 5 brings! :partying_face: 

I did a partial review so far, and it looks promising. However, I noticed Storybook is broken (`npm run storybook`), even though in `trunk` it worked fine. Can you reproduce as well?

**Comment by thealexandrelara** [comment]
> Thanks for working on this! Do you think that it is fine having `${number}-frontend`? I remember that I tried to fix it without success. On the PHP side, we register scripts with the structure `wc-${block_name}-block-frontend`.
> 
> image

Oh nice catch! You mean because some numeric files were created during the build? I would have to investigate why Webpack made those files after the upgrade. Can you share from which file you took this code (in your image)?

File: package.json
Comment (thealexandrelara) [reply package.json]: > Really exciting to see the bundle size improvements webpack 5 brings! 🥳

Yeahh, that's really awesome to see the bundle size decreasing that much 🎉

> I did a partial review so far, and it looks promising. However, I noticed Storybook is broken (`npm run storybook`), even though in `trunk` it worked fine. Can you reproduce as well?

Thank you for pointing this out! I just fixed that. By default, Storybook uses Webpack 4 to build everything, [so I tweaked its configurations](https://storybook.js.org/docs/react/builders/webpack#webpack-5) to make it use Webpack 5


> If I remember correctly, we use fixed versions in `package.json` (_aka_, without the `^` caret operator), because renovatebot already takes care of updating the dependencies.

Great, I just removed the caret operator from the versions there. Thank you!

**Comment by gigitux** [comment]
> Oh nice catch! You mean because some numeric files were created during the build? 

Yes! 

> I would have to investigate why Webpack made those files after the upgrade. 

👍 . Otherwise, we can check that this new filename doesn't impact the project. 

> Can you share from which file you took this code (in your image)?

There are multiple instances, search in the project with the keyword "-block-frontend" and you will find multiple results.

I took the screenshot from `Cart.php`.

**Comment by thealexandrelara** [comment]
> > Oh nice catch! You mean because some numeric files were created during the build?
> 
> Yes!
> 
> > I would have to investigate why Webpack made those files after the upgrade.
> 
> 👍 . Otherwise, we can check that this new filename doesn't impact the project.
> 
> > Can you share from which file you took this code (in your image)?
> 
> There are multiple instances, search in the project with the keyword "-block-frontend" and you will find multiple results.
> 
> I took the screenshot from `Cart.php`.

I've finally discovered what is actually happening. By default, Webpack 5 splits modules that are shared between multiple entries. I've found that the shared chunks were already happening before; however, chunk filenames were different because they were using the `splitChunks.automaticNameDelimiter` property in the Webpack config to name chunks. So, for example, if we had Cart and Checkout components sharing the same piece of code, it would result in a file `cart--checkout-frontend.js`.

When it comes to Webpack 5 when they generate shared chunks, [those chunks do not have a name associated with them](https://github.com/webpack/webpack/discussions/14451), but they have an id (that is numeric), so they use that `id` to name the file. Looking at the code, it seems the PHP files are just importing the entry files (i.e.: `cart-frontend.js`, `checkout-frontend.js`), and Webpack is responsible for looking at the chunk graph and importing the required chunks for that piece of code to work. It was the same before, but with a different naming schema, so I think this wouldn't have a negative impact on the project. We can keep testing and check if any weird behavior appears.

**Review by Aljullu — COMMENTED** [review]
Confirming Storybook works well now. I also did some smoke testing and most things work as expected, however, I noticed that on RTL languages, styles seem to be broken.

This is with the language set to Arabic:

`trunk` | This branch
--- | ---
imatge | imatge

Can you reproduce as well, @thealexandrelara?

**Comment by kmanijak** [comment]
I was checking the security vulnerability alerts and I think [this one](https://github.com/woocommerce/woocommerce-blocks/security/dependabot/16) will be possible fixed after updating Webpack to version 5.

**Comment by thealexandrelara** [comment]
> Confirming Storybook works well now. I also did some smoke testing and most things work as expected, however, I noticed that on RTL languages, styles seem to be broken.
>
> Can you reproduce as well, @thealexandrelara?

I can't reproduce it, but I made some changes to the Webpack config to generate css files with the suffix `-rtl.css` as [recommended by WordPress](https://codex.wordpress.org/Right-to-Left_Language_Support). The library that we are using (@automattic/webpack-rtl-plugin) generates files with the suffix `.rtl.css` and, unfortunately, does not provide an option to change that

File: bin/webpack-configs.js
Comment (Aljullu) [inline bin/webpack-configs.js]: Nit: I think it's better to always use multiline clauses. This will keep consistency with our codebase and IMO makes code easier to read.
```suggestion
							if ( ! module.type.includes( 'css' ) ) {
								return false;
							}
```

**Review by Aljullu — COMMENTED** [review]
Thanks for updating this PR, @thealexandrelara! This fixes the issue for me; however, I'm not convinced of having to maintain a copy of `webpack-rtl-plugin`. IMO it would be better to:

1. Try to upstream to `@automattic/webpack-rtl-plugin` the option to use dashes instead of dots (`-rtl.css` instead of `.rtl.css`).
2. Use `rtlcss` directly, in a similar fashion to [what Gutenberg is doing](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L128-L134).

What do you think? I'm a bit wary that having a copy of `webpack-rtl-plugin` will increase the amount of maintenance work we have to do in the future as we need to keep up with updates.

**Comment by thealexandrelara** [comment]
> Thanks for updating this PR, @thealexandrelara! This fixes the issue for me; however, I'm not convinced of having to maintain a copy of `webpack-rtl-plugin`. IMO it would be better to:
> 
> 1. Try to upstream to `@automattic/webpack-rtl-plugin` the option to use dashes instead of dots (`-rtl.css` instead of `.rtl.css`).
> 2. Use `rtlcss` directly, in a similar fashion to [what Gutenberg is doing](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L128-L134).
> 
> What do you think? I'm a bit wary that having a copy of `webpack-rtl-plugin` will increase the amount of maintenance work we have to do in the future as we need to keep up with updates.

Thank you for reviewing and for the suggestions! I think the first option would be preferable because some things concern me about the second option:
1. To make it work, Gutenberg is making use of Workers to manipulate the files during the build. They have to manually [render the SASS files](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L115), process the result  using postcss, and also manually manipulate the files ([retrieving](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L84), [changing the filenames ](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L85)and [saving them](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L137));
2. They also make use of a `stream` object that stores and processes the files, including being responsible for building the css;
3. They are using the `node-worker-farm` lib to create the worker responsible for processing the css, but that library is not being maintained anymore;

It seems most of the code is from 2-4 years ago, but since then, Webpack has evolved, and we can probably benefit from some new built-in resources without having to manually manipulate and process the files. 

Considering that, I think the best option would be to upstream the changes to `@automattic/webpack-rtl-plugin`. One thing we should consider is if we have to block this card until they release a new version that allows us to change the filename extension or if we can use this temporary solution (with our version of /the Webpack RTL Plugin) and open a new issue to replace this solution with the latest version in the future. What do you think?

**Review by Aljullu — APPROVED** [review]
> Considering that, I think the best option would be to upstream the changes to `@automattic/webpack-rtl-plugin`. One thing we should consider is if we have to block this card until they release a new version that allows us to change the filename extension or if we can use this temporary solution (with our version of /the Webpack RTL Plugin) and open a new issue to replace this solution with the latest version in the future. What do you think?

That's a good point. Probably no need to block updating to Webpack 5 because of this. Will you open an issue about this, though? This way we don't forget. :slightly_smiling_face: 

I did another round of testing and everything is looking great on my end. Good work, @thealexandrelara! :ship:

**Comment by thealexandrelara** [comment]
> Will you open an issue about this, though? This way we don't forget. 🙂
> 
> I did another round of testing and everything is looking great on my end. Good work, @thealexandrelara! 🚢

Sure I can open an issue about this! Thank you 🙌

**Comment by thealexandrelara** [comment]
Just providing a quick update on why this PR was not merged yet: I was reviewing the Github checks and noticed that some unit tests are broken on this branch, probably caused by a babel configuration. I'm currently working on other issues that were prioritized, but I'll be working on this one again as soon as I can

**Comment by ralucaStan** [comment]
@thealexandrelara I am converting this to draft. Feel free to remove this once this PR is opened for review

**Comment by opr** [comment]
Removed the milestone - please re-add it when this PR gets merged.

**Review by nefeline — COMMENTED** [review]
Hey @thealexandrelara ! I'm noticing some oddities in the editor when inserting certain blocks or patterns. 

For example, the filter by attribute, filter by stock status, and the filter by rating blocks have all display issues in the editor:

Screenshot 2023-08-10 at 09 22 41

Are you able to reproduce this as well?

**Comment by tjcafferkey** [comment]
>For example, the filter by attribute, filter by stock status, and the filter by rating blocks have all display issues in the editor:

@nefeline I ran a clean install and build and I'm not encountering this problem. Just wondering whether you've done the same?

Below is a snippet of my zsh command, and a screenshot of the filter blocks in editor.

```
function wcb-fresh-install {
  cd ~/LocalSites/woodevelopment/app/public/wp-content/plugins/woocommerce-blocks
  nvm-install
  rm -rf node_modules
  composer install
  npm install
  npm run build
  npm run start
}
```

Screenshot 2023-08-10 at 08 56 36

**Comment by nefeline** [comment]
> I ran a clean install and build and I'm not encountering this problem. Just wondering whether you've done the same?

I ran the following:

- nvm use
- rm -rf node_modules
- npm install
- npm run build

Composer should not be interfering with this webpack update, but I also ran it on a second test just in case, I'm getting the same results:

Screen Recording 2023-08-10 at 11 11 33

**Comment by thealexandrelara** [comment]
> Hey @thealexandrelara ! I'm noticing some oddities in the editor when inserting certain blocks or patterns.
> 
> For example, the filter by attribute, filter by stock status, and the filter by rating blocks have all display issues in the editor:

> Are you able to reproduce this as well?

Hey @nefeline, thank you for reporting this, I was able to reproduce this as well and just pushed a fix to it. Let me know if this solves the issue for you too

**Review by nefeline — APPROVED** [review]
> Hey @nefeline, thank you for reporting this, I was able to reproduce this as well and just pushed a fix to it. Let me know if this solves the issue for you too

Thanks for the update @thealexandrelara! I can confirm https://github.com/woocommerce/woocommerce-blocks/pull/8013/commits/426670e9d5cdf2069904e3ad8b25ee130b11020f solves the problem.

I ran a series of tests with the following blocks:

* All filter blocks: filter by stock, filter by attribute, filter by rating, filter by price
* Active Filters
* Products (Beta)
* Single Product
* Products by Category
* Products by Attribute
* Mini-cart
* Product Search
* Product tags
* Featured Product

They all work as expected, and the styles are correct when running both on development and production builds. 

Can we ensure this branch is up-to-date with trunk and solve the conflicts, please? I'll do another pass & re-review as soon as everything is up-to-date.

I'm pre-approving as the styles are all correct, and blocks are fully functional now: but we should wait for others to finish their tests as planned on pdToLP-Ll-p2 before merging.

Great work! 🎉

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/index.tsx
Comment (nefeline) [inline assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/index.tsx]: Should we ditch these changes? Any particular reason for us to be removing `isExperimentalBuild` here?

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx
Comment (nefeline) [inline assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx]: Same as the previous: should we move this to a separate PR?

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/index.tsx
Comment (thealexandrelara) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/index.tsx]: The Product Gallery block contains some inner blocks. These inner blocks are only available inside the Product Gallery block, so even if I remove this, they will only be made available on the editor or frontend with its main block. 

The reason why I removed this is because of an issue that was occurring when using the `isExperimentalBuild()` (p1691416552602699/1691413062.084399-slack-C02UBB1EPEF) inside the inner blocks. I don't know exactly why this was happening, but I suspect the cause is that for the inner blocks to work, we have to import them in the `index.tsx` file of the [Product Gallery block](https://github.com/woocommerce/woocommerce-blocks/blob/chore/upgrade-webpack-to-version-5/assets/js/blocks/product-gallery/index.tsx#L14-L15), since there are many changes/improvements on the tree-shaken process in Webpack V5, sometimes Webpack can fail to recognize a file/piece of code that should be included in the bundle, in this particular case, it failed to include the `feature_flags` function that the `isExperimentalBuild()` call uses.

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx
Comment (thealexandrelara) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx]: We can try solving this in a separate PR if we consider it necessary. The only thing we need to have in mind is that this is an inner block of the Product Gallery block that is currently set to be experimental, which means that all its inner blocks will only be made available when used inside the Product Gallery block and only if this block is also available.

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/index.tsx
Comment (nefeline) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/index.tsx]: > The Product Gallery block contains some inner blocks. These inner blocks are only available inside the Product Gallery block, so even if I remove this, they will only be made available on the editor or frontend with its main block.

Gotcha, thanks for clarifying.

> The reason why I removed this is because of an issue that was occurring when using the isExperimentalBuild() (p1691416552602699/1691413062.084399-slack-C02UBB1EPEF) inside the inner blocks. I don't know exactly why this was happening, but I suspect the cause is that for the inner blocks to work, we have to import them in the index.tsx file of the [Product Gallery block](https://github.com/woocommerce/woocommerce-blocks/blob/chore/upgrade-webpack-to-version-5/assets/js/blocks/product-gallery/index.tsx#L14-L15), since there are many changes/improvements on the tree-shaken process in Webpack V5, sometimes Webpack can fail to recognize a file/piece of code that should be included in the bundle, in this particular case, it failed to include the feature_flags function that the isExperimentalBuild() call uses.

👍 thanks for the detailed explanation! If this only affects inner blocks and won't cause any problems for the `product-gallery` or the `product-collection` blocks (these are the ones currently behind the experimental flag), all good.

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx
Comment (nefeline) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx]: > We can try solving this in a separate PR if we consider it necessary. The only thing we need to have in mind is that this is an inner block of the Product Gallery block that is currently set to be experimental, which means that all its inner blocks will only be made available when used inside the Product Gallery block and only if this block is also available.

All good, thanks for the explanation on https://github.com/woocommerce/woocommerce-blocks/pull/8013#discussion_r1291708015 : I agree it makes sense to keep this change (but only if this limitation with the `isExperimentalBuild()` is exclusively affecting inner blocks and won't expose the gallery and product collection blocks to production ahead of time).

**Comment by nefeline** [comment]
@thealexandrelara I ran a second round of tests here and confirmed blocks are working as expected & styles are correct: thanks again for your prompt response & work here! 🙌! Let's see if anyone has any other findings based on the remaining tests: pdToLP-Ll-p2

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx
Comment (nefeline) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx]: BTW: do we have an issue opened to keep track of this limitation with the `isExperimentalBuild()` on inner blocks?

Comment (thealexandrelara) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx]: As of now, we don't have an issue opened specifically for this, but it's something worth considering. What would be expected from an issue regarding this limitation? Would the focus be on documenting this as a limitation, advising developers against utilizing isExperimentalBuild for inner blocks when the parent block is already using it? Or, would the idea be to find a way of enabling the use of multiple isExperimentalBuild instances in the parent block and its inner blocks?

Comment (nefeline) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx]: > What would be expected from an issue regarding this limitation?

IMO it should work independent of the block type:  `isExperimentalBuild()` should work as expected for both blocks and inner blocks (as it used to work before this Webpack update).

**Review by gigitux — APPROVED** [review]
I verified that E2E test env setup works correctly locally! LGTM!

**Review by opr — APPROVED** [review]
As mentioned in pdToLP-Ll-p2#comment-542 I have tested the Cart and Checkout blocks and could not find any issues!

**Comment by alexflorisca** [comment]
I've ran the build and start commands with the new webpack config and all seems fine for me too. The only thing I wanted to point out is that several of the bundles scripts have increased significantly in size. See the  bundle size https://github.com/woocommerce/woocommerce-blocks/pull/8013#issuecomment-1362082382 above - look for any with a 🆘 or 🚨 emojis. Have we addressed this?

I also did a quick test on the Cart page to see what the overall increase in assets is

| Cart Page 	| Webpack 4 	| Webpack 5 	|
|-----------	|-----------	|-----------	|
| JS        	| 4.1MB     	| 4.1MB     	|
| CSS       	| 335Kb     	| 351Kb     	|

The JS looks fine overall but the CSS has increased slightly and I wonder why that is? I also wonder if we should do this for other blocks, or maybe one page with all the blocks on it to see if there's any significant impact?

Think this is quite important to address before we merge as we don't want the bundle size to increase

**Comment by dinhtungdu** [comment]
@thealexandrelara I got this error while testing storybook against this PR:

```sh
**99% done plugins webpack-hot-middleware**webpack built preview 078da4b21ec586873670 in 1732ms
```

The script stucks at 99%. Viewing the storybook on the frontend, I only see the spinner running forever

image


I see you attempted to fix the storybook in https://github.com/woocommerce/woocommerce-blocks/tree/a82c20da5881d2b957f6a609e532b7a6d5a3cb51, I wonder if it's still working for you now.

**Comment by thealexandrelara** [comment]
> @thealexandrelara I got this error while testing storybook against this PR:
> 
> ```shell
> **99% done plugins webpack-hot-middleware**webpack built preview 078da4b21ec586873670 in 1732ms
> ```
> 
> The script stucks at 99%. Viewing the storybook on the frontend, I only see the spinner running forever
> 
> image
> I see you attempted to fix the storybook in https://github.com/woocommerce/woocommerce-blocks/tree/a82c20da5881d2b957f6a609e532b7a6d5a3cb51, I wonder if it's still working for you now.

You're right, I just tested it and can reproduce the issue as well. I attempted to resolve it by cleaning the Storybook cache by running `rm -rf node_modules/.cache/storybook`, but unfortunately, it didn't fix the issue.

I'll need to investigate what's causing this problem, given that it was working before. Perhaps upgrading to the latest version is the only option, although I was hoping to handle upgrades of external libraries through separate PRs.

**Comment by dinhtungdu** [comment]
@thealexandrelara I [tried](https://github.com/woocommerce/woocommerce-blocks/commit/b321aee83a280d5b602f4c2b39dadbeea3079cf5) using `storyStorev7` as Gutenberg is using but got new issues with some stories. I also tried [upgrading to Storybook v7](https://github.com/woocommerce/woocommerce-blocks/tree/try/storybook7) but got the same issues. I don't know how to fix those issues, so posting my trials here hoping it can help you or other fix them.

image

**Review by kmanijak — APPROVED** [review]
I checked the blocks if they render correctly and if they're functional.

For each of the blocks in [this list](https://wordpress.org/plugins/woo-gutenberg-products-block/) + Products (Beta) block:
-  ✅ add the block to the post before the upgrade. Upgrade WC Blocks to the version built with Webpack 5. No problems were found either in Editor or Frontend
-  ✅ add the block to the new post after the upgrade. No problems were found either in Editor or Frontend

**Review by nielslange — APPROVED** [review]
I've successfully tested the translations against the Webpack 5 update. There was only one minor translation issue, which will solve itself once this PR has been deployed to *.org and the GlotPress translations have been updated, see p1692365786355899/1692363111.353269-slack-C02UBB1EPEF.

**Review by imanish003 — APPROVED** [review]
Hello @thealexandrelara, great job! I've tested the Mini Cart block, and I'm pleased to report that it's functioning as anticipated. 🚀

**Comment by samueljseay** [comment]
@thealexandrelara @dinhtungdu I had a go at fixing the storybook issues today.

You can fix the immediate problem of unexpected token default,its a docgen issue, it doesn't like static class properties and mangles them.

Install `react-docgen-typescript-plugin` then add this to the main.js:

```
typescript: {
		reactDocgen: 'react-docgen-typescript-plugin',
},
```

(note that upgrading to storybook 7 I think also fixes this problem due to them updating the dep on docgen plugin, but if you wanted a way to avoid upgrading to storybook 7 then this is an option 😄 )

That does not however solve the `exports is not defined` error. That's coming from `rungen` which is a dep of `@wordpress/redux-routine`. I haven't been able to figure this one out yet, and also worth nothing that although storybook appears to display more correctly on `trunk`, it also, suffers from this exception so Storybook is already a bit broken.

**Comment by thealexandrelara** [comment]
Hey everyone,
Thank you for testing and approving the PR. Really great to see that we are almost there and hopefully soon we will be able to merge this PR 🙌 

> @thealexandrelara @dinhtungdu I had a go at fixing the storybook issues today.
> 
> You can fix the immediate problem of unexpected token default,its a docgen issue, it doesn't like static class properties and mangles them.
> 
> Install `react-docgen-typescript-plugin` then add this to the main.js:
> 
> ```
> typescript: {
> 		reactDocgen: 'react-docgen-typescript-plugin',
> },
> ```
@samueljseay Ohh thank you so much!!! I just pushed the changes as you suggested and the Storybook is working again!


> 
> (note that upgrading to storybook 7 I think also fixes this problem due to them updating the dep on docgen plugin, but if you wanted a way to avoid upgrading to storybook 7 then this is an option 😄 )
> 
> That does not however solve the `exports is not defined` error. That's coming from `rungen` which is a dep of `@wordpress/redux-routine`. I haven't been able to figure this one out yet, and also worth nothing that although storybook appears to display more correctly on `trunk`, it also, suffers from this exception so Storybook is already a bit broken.

Thank you for investigating this, in this case, I believe we should address this problem in a separate PR, probably we can create an issue to upgrade Storybook and work on a fix to this since it is already broken on `trunk`

**Review by wavvves — APPROVED** [review]
I tested build, and cart/checkout templates local built branch and via zip, and spotted no different behaviour.

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx
Comment (thealexandrelara) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx]: @nefeline After adding the `"./assets/js/settings/blocks/index.ts"` file to the `sideEffects` property of the `package.json` file, the issue is no longer happening, so I restored these files to how it was before 🙌

**Review by dinhtungdu — APPROVED** [review]
Storybook is testing great with the latest commit. 🎉 Thank you so much for the great work @thealexandrelara and @samueljseay!

`ProductPrice` store has an error but the current `trunk` storybook has that same error too, so it shouldn't block this PR.
image

**Comment by tarhi-saad** [comment]
@thealexandrelara, I bumped this PR to the next release!

**Pull request merged**
